src/transport/raw/TCP.cpp

```diff
@@ -359,126 +422,171 @@ CHIP_ERROR TCPBase::OnTcpReceive(Inet::TCPEndPoint * endPoint, System::PacketBuf
-            System::PacketBufferHandle buffer = std::move(pending->mPacketBuffer);
-            tcp->mPendingPackets.ReleaseObject(pending);
+            tcp->CloseConnectionInternal(activeConnection, err, SuppressCallback::No);
+            return;
+        }
 
-            if ((inetErr == CHIP_NO_ERROR) && (err == CHIP_NO_ERROR))
+        // Send any pending packets that are queued for this connection
+        tcp->mPendingPackets.ForEachActiveObject([&](PendingPacket * pending) {
+            if (pending->mPeerAddress == addr)
             {
-                err = endPoint->Send(std::move(buffer));
+                foundPendingPacket                = true;
+                System::PacketBufferHandle buffer = std::move(pending->mPacketBuffer);
+                tcp->mPendingPackets.ReleaseObject(pending);
+
+                if ((conErr == CHIP_NO_ERROR) && (err == CHIP_NO_ERROR))
+                {
+                    err = endPoint->Send(std::move(buffer));
+                }
             }
-        }
-        return Loop::Continue;
-    });
+            return Loop::Continue;
+        });
 
-    if (err == CHIP_NO_ERROR)
-    {
-        err = inetErr;
-    }
+        // Set the TCPKeepalive configurations on the established connection
+        endPoint->EnableKeepAlive(activeConnection->mTCPKeepAliveIntervalSecs, activeConnection->mTCPMaxNumKeepAliveProbes);
 
-    if (!foundPendingPacket && (err == CHIP_NO_ERROR))
-    {
-        // Force a close: new connections are only expected when a
-        // new buffer is being sent.
-        ChipLogError(Inet, "Connection accepted without pending buffers");
-        err = CHIP_ERROR_CONNECTION_CLOSED_UNEXPECTEDLY;
-    }
+        ChipLogProgress(Inet, "Connection established successfully with [%s:%d]", ipAddrStr, port);
 
-    // cleanup packets or mark as free
-    if (err != CHIP_NO_ERROR)
-    {
-        ChipLogError(Inet, "Connection complete encountered an error: %s", ErrorStr(err));
-        endPoint->Free();
-        tcp->mUsedEndPointCount--;
+        // Let higher layer/delegate know that connection is successfully
+        // established
+        tcp->HandleConnectionComplete(activeConnection, CHIP_NO_ERROR);
     }
     else
     {
-        bool connectionStored = false;
-        for (size_t i = 0; i < tcp->mActiveConnectionsSize; i++)
-        {
-            if (!tcp->mActiveConnections[i].InUse())
-            {
-                tcp->mActiveConnections[i].Init(endPoint);
-                connectionStored = true;
-                break;
-            }
-        }
-
-        // since we track end points counts, we always expect to store the
-        // connection.
-        if (!connectionStored)
-        {
-            endPoint->Free();
-            ChipLogError(Inet, "Internal logic error: insufficient space to store active connection");
-        }
+        ChipLogError(Inet, "Connection establishment with [%s:%d] encountered an error: %s", ipAddrStr, port, ErrorStr(err));
```

Comment:
Please use CHIP_ERROR_FORMAT instead of forcing strings.

Reply:
Done.